### PR TITLE
#10246 Add access quality definitions to cluster definitions

### PR DESCRIPTION
**Pull request by hubTab**
#### Problem
"Access" privileges are not defined yet in any cluster XML.

#### Change overview
Annotated the operational-credentials-cluster.xml and onoff-cluster.xml based on the Matter Specs, section 11.18, and appclusters.pdf (locally built), respectively.

#### Testing
- Regenerated ZAP and did not fail. No templates use these yet.

### CODE REVIEW

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Please actually fill out the PR template (what's being changed, why, and how it was tested).

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml [hunk 1]
```diff
@@ -54,43 +54,65 @@ limitations under the License.
...
+      <description>CommissionedFabrics</description>
+      <access op="read" role="view"/>
+    </attribute>
     <!-- 400 = 400 bytes for root cert -->
-    <attribute side="server" code="0x0004" define="TRUSTED_ROOTS" type="ARRAY" entryType="OCTET_STRING" length="400" writable="false" optional="false">TrustedRootCertificates</attribute>
+    <attribute side="server" code="0x0004" define="TRUSTED_ROOTS" type="ARRAY" entryType="OCTET_STRING" length="400" writable="false" optional="false">
+      <description>TrustedRootCertificates</description>
+      <access op="read" role="view"/>
+    </attribute>
+    <attribute side="server" code="0x0005" define="CURRENT_FABRIC" type="INT8U" default="0" writable="false" optional="false">
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:74]: This needs merging to tip anyway, where this attribute already exists...

File: src/app/zap-templates/zcl/data-model/chip/onoff-cluster.xml [hunk 2]
```diff
@@ -48,39 +48,154 @@ limitations under the License.
...
+      | ID     | Name               | Type       | Constraint  | Quality | Default | Access | Conformance
+      | 0x0000 | OnOff              | bool       | all         | SN      | FALSE   | R  V   | M
+
+      <access/> Design Rationale:
+        Per Definition:
+          Operation:        R(ead) only. I.e. Write is inferred as not allowed.
+          Role/Privilege:   V(iew)
+          Modifier:         None
+    -->
+    <attribute side="server" code="0x0000" define="ON_OFF" type="boolean" default="0" reportable="true" optional="false">
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/onoff-cluster.xml:62]:
```suggestion
    <attribute side="server" code="0x0000" define="ON_OFF" type="boolean" default="0"  writable="false" reportable="true" optional="false">
```

File: src/app/zap-templates/zcl/data-model/chip/onoff-cluster.xml [hunk 3]
```diff
@@ -48,39 +48,154 @@ limitations under the License.
...
+          Operation:        R(ead) 
+                            W(rite)
+          Role/Privilege:   V(iew) for Read
+                            O(perate) for Write
+          Modifier:         None
+    -->
+    <attribute side="server" code="0x4001" define="ON_TIME" type="int16u" writable="true" optional="true">
+      <description>OnTime</description>
+      <access op="read" privilege="view"/>
+      <access op="write" privilege="operate"/>
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/onoff-cluster.xml:99]:
```suggestion
      <description>OnTime</description>
      <!-- The following required access privileges are the default, and could have been
            omitted, but are included here as an example. -->
      <access op="read" privilege="view"/>
      <access op="write" privilege="operate"/>
```

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml [hunk 4]
```diff
@@ -54,43 +54,187 @@ limitations under the License.
...
+          Operation:        R(ead) only. I.e. Write is inferred as not allowed.
+          Role/Privilege:   A(dminister)
+          Modifier:         F(abric-scoped)
+
+      Other notes:
+        Newly added attribute, as per the spec, as it was not included.
+    -->
+    <attribute side="server" code="0x0000" define="NOCs" type="ARRAY" entryType="NOCStruct" length="320" writable="false" optional="false">
+      <description>NOCs</description>
+      <access op="read" privilege="administer" modifier="fabric-scoped"/>
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:85]:
```suggestion
      <access op="read" privilege="administer"/>
      <access modifier="fabric-scoped"/>
```

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml [hunk 5]
```diff
@@ -54,43 +54,187 @@ limitations under the License.
...
+          Operation:        R(ead) only. I.e. Write is inferred as not allowed.
+          Role/Privilege:   V(iew)
+          Modifier:         F(abric-scoped)
+
+      Other notes:
+        Updated attribute description, as per the spec.
+    -->
+    <attribute side="server" code="0x0001" define="FABRICS" type="ARRAY" entryType="FabricDescriptor" length="320" writable="false" optional="false">
+      <description>Fabrics</description>
+      <access op="read" privilege="view" modifier="fabric-scoped"/>
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:104]:
```suggestion
      <access op="read" privilege="view"/>
      <access modifier="fabric-scoped"/>
```

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml [hunk 6]
```diff
@@ -54,43 +54,187 @@ limitations under the License.
...
+
+      <access/> Design Rationale:
+        Per Definition:
+          Operation:        R(ead) only. I.e. Write is inferred as not allowed.
+          Role/Privilege:   V(iew)
+          Modifier:         None
+    -->    
     <!-- 400 = 400 bytes for root cert -->
-    <attribute side="server" code="0x0004" define="TRUSTED_ROOTS" type="ARRAY" entryType="OCTET_STRING" length="400" writable="false" optional="false">TrustedRootCertificates</attribute>
+    <attribute side="server" code="0x0004" define="TRUSTED_ROOTS" type="ARRAY" entryType="OCTET_STRING" length="400" writable="false" optional="false">
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:151]: Each ENTRY can be 400. Will 400 limit be a problem, or is it ignored?

@bzbarsky-apple

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml [hunk 7]
```diff
@@ -54,43 +54,187 @@ limitations under the License.
...
+      | Id | Name                | Type                                      | Constraint            | Quality | Default | Access | Conformance
+      | 5  | CurrentFabricIndex  | uint8                                     |                       |         | 0       | R V    | M
+
+      <access/> Design Rationale:
+        Per Definition:
+          Operation:        R(ead) only. I.e. Write is inferred as not allowed.
+          Role/Privilege:   V(iew)
+          Modifier:         None
+    -->    
+    <attribute side="server" code="0x0005" define="CURRENT_FABRIC_INDEX" type="fabric_idx" writable="false" optional="false">
```
Comment (tcarmelveilleux) [inline src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:167]: Uses type `fabric_idx`. We need to fix #11281

File: src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml (hunk 6 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/operational-credentials-cluster.xml:151]: At this point, I believe the `length` is completely ignored; it used to be used for allocating space in the attr store, but we don't do that anymore.

Just looking through this quickly, in the ZAP repo we have:
```
% git grep "\.\$\.length"
src-electron/zcl/zcl-loader-silabs.js:        maxLength: attribute.$.length ? attribute.$.length : null,
src-electron/zcl/zcl-loader-silabs.js:        maxLength: item.$.length ? item.$.length : null,
```
and then that's used in `typeSizeAttribute`.  `typeSizeAttribute` is used in `helper-endpointconfig.js` and `helper-tokens.js`, neither of which is relevant to list attributes in Matter.  And in Matter we use `maxLength` directly in `Accessors-src.zapt`, which is also not relevant to list attributes.

Not sure whether we should just drop it from our XML or what.  @vivien-apple any opinions?

**Pull request closed without merging**